vision.py

```diff
@@ -33,106 +33,123 @@ def run(self):
+    best_group = rank_groups(groups, contour_areas)
+    if best_group is None:
+        display = annotate_image(frame, contours, [], (-1, -1))
+        return [None, display]
+    values = get_values(contours, best_group, (frame.shape[1], frame.shape[0]), contour_areas)
+    display = annotate_image(frame, contours, best_group, values)
+    return values, display
+
+def preprocess(frame: np.ndarray) -> np.ndarray:
+    """Creates a mask of expected target green color
+    """
+    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
+    mask = cv2.inRange(hsv, TARGET_HSV_LOW, TARGET_HSV_HIGH)
+    mask = cv2.erode(mask, ERODE_DILATE_KERNEL)
+    mask = cv2.dilate(mask, ERODE_DILATE_KERNEL)
+    return mask
+
+def find_contours(mask: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    """Finds contours on a grayscale mask
+    """
+    *_, contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+    return contours
+
+def group_contours(contours: np.ndarray, contour_areas: list) -> List[List[int]]:
+    """Returs a nested list of contour indices grouped by relative distance
+    """
+    # Build an adjacency list based on distance between contours relative to their area
+    connections = [[] for _ in contours]
+    for i, a in enumerate(contours):
+        a_com = a.mean(axis=0)[0]
+        a_area = contour_areas[i]
+        for j, b in enumerate(contours[i + 1:]):
+            max_metric = max(a_area, contour_areas[j+i+1]) * METRIC_SCALE_FACTOR
+            b_com = b.mean(axis=0)[0]
+            d = a_com - b_com
+            metric = d[0] ** 2 + d[1] ** 2
+            if (metric < max_metric):
+                connections[i].append(j + i + 1)
+                connections[j + i + 1].append(i)
+    # Breadth first search from each contour that wasn't yet assigned to a group to find its group
+    assigned = [False for _ in contours]
+    group = set()
+    groups = []
+    for i, c in enumerate(connections):
+        if assigned[i]: continue
+        group = {i}
+        assigned[i] = True
+        horizon = set(c)
+        while (True):
+            if len(horizon) == 0:
+                # Group completed
+                groups.append(list(group))
+                break
+            for i in horizon: assigned[i] = True
+            group.update(horizon)
+            new_horizon = set()
+            for old in horizon:
+                new_horizon.update(connections[old])
+            new_horizon.difference_update(group)
+            horizon = new_horizon
+    return groups
+
+def rank_groups(groups: List[List[int]], contour_areas: np.ndarray) -> Optional[List[int]]:
+    """Returns the group that is most likely to be the target
+    Could use metrics such as curvature, consistency of contours, consistency with last frame
+
+    Currently just returns the group with most elements
+    """
+    return max((g for g in groups if len(g) > 1), key=lambda x: sum(contour_areas[i] for i in x), default=None)
```

Comment:
Why throw away the groups if they're a group of only 1? (Please add this in a code comment.)